src/workerd/io/actor-sqlite.c++

```diff
@@ -169,11 +200,67 @@ void ActorSqlite::onWrite() {
       // rather than after the callback.
       { auto drop = kj::mv(txn); }
 
-      return commitCallback();
+      return commitImpl(kj::mv(precommitAlarmPromise));
     })));
   }
 }
 
+kj::Promise<void> ActorSqlite::requestScheduledAlarm(kj::Maybe<kj::Date> requestedTime) {
+  co_await hooks.scheduleRun(requestedTime);
+  lastConfirmedScheduledAlarm = requestedTime;
+}
+
+ActorSqlite::PrecommitAlarmState ActorSqlite::startPrecommitAlarmScheduling() {
+  PrecommitAlarmState state;
+  state.localAlarmState = metadata.getAlarm();
+  if (pendingCommit == kj::none &&
+      willFireEarlier(state.localAlarmState, lastConfirmedScheduledAlarm)) {
+    // Basically, this is the first scheduling request that commitImpl() would make prior to
+    // commitCallback().  We start the request separately, ahead of calling sqlite functions that
+    // commit to local disk, for correctness in workerd, where alarm scheduling and db commits are
+    // both synchronous.
+    state.schedulingPromise = requestScheduledAlarm(state.localAlarmState);
+  }
+  return kj::mv(state);
+}
+
+kj::Promise<void> ActorSqlite::commitImpl(ActorSqlite::PrecommitAlarmState precommitAlarmState) {
+  // We assume that exceptions thrown during commit will propagate to the caller, such that they
+  // will ensure cancelDeferredAlarmDeletion() is called, if necessary.
+
+  KJ_IF_SOME(pending, pendingCommit) {
+    co_await pending.addBranch();
+    co_return;
+  }
+  auto [promise, fulfiller] = kj::newPromiseAndFulfiller<void>();
+  pendingCommit = promise.fork();
+
+  // Wait for any precommit alarm scheduling work to complete.
+  auto localAlarmState = precommitAlarmState.localAlarmState;
+  KJ_IF_SOME(p, precommitAlarmState.schedulingPromise) {
+    co_await p;
+    localAlarmState = metadata.getAlarm();
+  }
+
+  while (willFireEarlier(localAlarmState, lastConfirmedScheduledAlarm)) {
+    co_await requestScheduledAlarm(localAlarmState);
+    localAlarmState = metadata.getAlarm();
+  }
+
+  auto commitCallbackPromise = commitCallback();
+
+  // Fulfill and clear the pending commit.
```

Comment:
What happens if code above throws? I think we need to reject the fulfiller on the way out so that other callers can't end up hung? (We can keep `pendingCommit` set, though... so all future commits will also throw, which seems fine, because any commit throwing is going to break the output gate anyway.)

Reply:
Am I mistaken in thinking that if the code throws, the fulfiller would be destroyed, which should reject the promise and unblock any caller waiting on the forked promise?

But I suppose in that case it will have the PromiseFulfiller exception message instead of the thrown exception, which would be more useful.  I can try adding that.

Reply:
Oh, interesting... Apparently we have logic in PromiseFulfiller to detect when it is being destroyed during unwind for a thrown exception and use that instead of "PromiseFulfiller destroyed", so we're even currently recording a copy of the thrown exception correctly.

https://github.com/capnproto/capnproto/blob/7ad12031/c%2B%2B/src/kj/async-test.c%2B%2B#L417-L429

I can add a test case.

Reply:
Oh... I forgot about that. Probably worth a comment.

I knew the fulfiller would be automatically rejected but thought it would be with an unhelpful "PromiseFulfiller was destroyed without fulfilling the promise" message.

Reply:
Added comment in d1983124b .